File: keyway/src/domains/access/transport/http.rs
//! Grants over HTTP.
//!
//! Delegating is scriptable and destroying is not (ADR-0005): a mistaken grant
//! is visible in the audit log and revocable in a click, whereas a deleted
//! secret has no undo.

use crate::AppState;
use crate::domains::access::entity::{Basis, Delegation, Level, Subject};
use crate::domains::audit::entity::{Action, Record};
use crate::domains::secrets::entity::identity;
use crate::transport::{ApiError, Caller};
use axum::extract::{Path, State};
use axum::routing::get;
use axum::{Json, Router};
use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/secrets/{id}/grants", get(list).post(delegate))
        .route(
            "/secrets/{id}/grants/{grant}",
            axum::routing::delete(revoke),
        )
}

#[derive(Serialize)]
struct GrantView {
    id: Uuid,
    subject_kind: String,
    subject: String,
    level: Level,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    keys: Vec<String>,
    granted_by: String,
    granted_at: DateTime<Utc>,
    expires_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "String::is_empty")]
    note: String,
}

impl From<Delegation> for GrantView {
    fn from(grant: Delegation) -> Self {
        Self {
            id: grant.id,
            subject_kind: grant.subject.kind().to_owned(),
            subject: grant.subject.id().to_owned(),
            level: grant.level,
            keys: grant.keys,
            granted_by: grant.granted_by,
            granted_at: grant.granted_at,
            expires_at: grant.expires_at,
            note: grant.note,
        }
    }
}

/// Who can see this secret — the list the whole mechanism exists to make
/// readable.
async fn list(
    State(state): State<AppState>,
    Caller(actor): Caller,
    Path(id): Path<Uuid>,
) -> Result<Json<Vec<GrantView>>, ApiError> {
    let (store, secret) = locate(&state, &actor, id).await?;
    Ok(Json(
        state
            .access
            .grants_on(&store, &secret)
            .await
            .map_err(ApiError::Internal)?
            .into_iter()
            .map(Into::into)
            .collect(),
    ))
}

#[derive(Deserialize)]
struct DelegateBody {
    /// `user` or `group`, always explicit — a team called `sre` and a person
    /// called `sre` are different subjects (ADR-0003).
    subject_kind: String,
    subject: String,
    level: String,
    #[serde(default)]
    keys: Vec<String>,
    #[serde(default)]
    days: i64,
    #[serde(default)]
    note: String,
}

async fn delegate(
    State(state): State<AppState>,
    Caller(actor): Caller,
    Path(id): Path<Uuid>,
    Json(body): Json<DelegateBody>,
) -> Result<Json<GrantView>, ApiError> {
    let (store, secret) = locate(&state, &actor, id).await?;

    // Only an owner or an admin hands out sight of a secret. A delegation at
    // `write` does not carry the right to re-delegate: that belongs to
    // ownership, which is a different act with a different audit line.
    let access = state
        .access
        .access_for(&actor, &store, &secret, Utc::now())
        .await
        .map_err(ApiError::Internal)?;
    if !matches!(access.basis, Basis::Owner | Basis::Admin) {
        return Err(ApiError::Forbidden);
    }

    let subject = match body.subject_kind.as_str() {
        "user" => Subject::User(body.subject.clone()),
        "group" => Subject::Group(body.subject.clone()),
        other => {
            return Err(ApiError::BadRequest(format!(
                "subject_kind must be user or group, not {other:?}"
            )));
        }
    };
    let level: Level =
        body.level
            .parse()
            .map_err(|e: crate::domains::access::entity::UnknownLevel| {
                ApiError::BadRequest(e.to_string())
            })?;

    let grant = Delegation {
        id: Uuid::new_v4(),
        store: store.clone(),
        secret: secret.clone(),
        subject,
        level,
        keys: body.keys,
        granted_by: actor.handle().to_owned(),
        granted_at: Utc::now(),
        expires_at: (body.days > 0).then(|| Utc::now() + Duration::days(body.days)),
        note: body.note,
    };

    state
        .access
        .delegate(&grant)
        .await
        .map_err(ApiError::Internal)?;

    state
        .audit
        .record(
            &actor,
            Record::new(Action::Delegate, id, &store, &secret)
                .subject(&body.subject)
                .keys(grant.keys.clone())
                .note(&grant.note),
        )
        .await
        .map_err(ApiError::Internal)?;

    Ok(Json(grant.into()))
}

async fn revoke(
    State(state): State<AppState>,
    Caller(actor): Caller,
    Path((id, grant_id)): Path<(Uuid, Uuid)>,
) -> Result<axum::http::StatusCode, ApiError> {
    let (store, secret) = locate(&state, &actor, id).await?;

    let access = state
        .access
        .access_for(&actor, &store, &secret, Utc::now())
        .await
        .map_err(ApiError::Internal)?;
    if !matches!(access.basis, Basis::Owner | Basis::Admin) {
        return Err(ApiError::Forbidden);
    }

    let subject = state
        .access
        .grants_on(&store, &secret)
        .await
        .map_err(ApiError::Internal)?
        .into_iter()
        .find(|g| g.id == grant_id)
        .map(|g| g.subject.id().to_owned())
        .ok_or(ApiError::NotFound)?;

    if !state
        .access
        .revoke(grant_id)
        .await
        .map_err(ApiError::Internal)?
    {
        return Err(ApiError::NotFound);
    }

    state
        .audit
        .record(
            &actor,
            Record::new(Action::Revoke, id, &store, &secret).subject(&subject),
        )
        .await
        .map_err(ApiError::Internal)?;

    Ok(axum::http::StatusCode::NO_CONTENT)
}

/// Which secret a uuid names, for a caller who can already see it.
async fn locate(
    state: &AppState,
    actor: &crate::domains::identity::entity::Actor,
    id: Uuid,
) -> Result<(String, String), ApiError> {
    let now = Utc::now();
    for store in state.stores.all() {
        let Ok(secrets) = store.list().await else {
            continue;
        };
        let Some(secret) = secrets.into_iter().find(|s| identity::id_for(s) == id) else {
            continue;
        };
        let access = state
            .access
            .access_for(actor, &secret.store, &secret.name, now)
            .await
            .map_err(ApiError::Internal)?;
        if !access.is_visible() {
            return Err(ApiError::NotFound);
        }
        return Ok((secret.store, secret.name));
    }
    Err(ApiError::NotFound)
}

File: internal/domains/access/transport/http.go
// Package transport is grants over HTTP.
//
// Delegating is scriptable and destroying is not (ADR-0005): a mistaken grant
// is visible in the audit log and revocable in a click, whereas a deleted
// secret has no undo.
package transport

import (
	"fmt"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/kotsmile/keyway/internal/domains/access/entity"
	auditentity "github.com/kotsmile/keyway/internal/domains/audit/entity"
	identityentity "github.com/kotsmile/keyway/internal/domains/identity/entity"
	secretsentity "github.com/kotsmile/keyway/internal/domains/secrets/entity"
	"github.com/kotsmile/keyway/internal/transport"
)

// Mount registers the grant routes on the authenticated API router.
func Mount(api chi.Router, state *transport.State) {
	api.Get("/secrets/{id}/grants", transport.Handler(list(state)))
	api.Post("/secrets/{id}/grants", transport.Handler(delegate(state)))
	api.Delete("/secrets/{id}/grants/{grant}", transport.Handler(revoke(state)))
}

// GrantView is a delegation as the API reports it.
type GrantView struct {
	ID          uuid.UUID    `json:"id"`
	SubjectKind string       `json:"subject_kind"`
	Subject     string       `json:"subject"`
	Level       entity.Level `json:"level"`
	Keys        []string     `json:"keys,omitempty"`
	GrantedBy   string       `json:"granted_by"`
	GrantedAt   time.Time    `json:"granted_at"`
	ExpiresAt   *time.Time   `json:"expires_at"`
	Note        string       `json:"note,omitempty"`
}

func viewOf(grant entity.Delegation) GrantView {
	return GrantView{
		ID:          grant.ID,
		SubjectKind: grant.Subject.Kind(),
		Subject:     grant.Subject.ID(),
		Level:       grant.Level,
		Keys:        grant.Keys,
		GrantedBy:   grant.GrantedBy,
		GrantedAt:   grant.GrantedAt,
		ExpiresAt:   grant.ExpiresAt,
		Note:        grant.Note,
	}
}

// list is who can see this secret — the list the whole mechanism exists to
// make readable.
func list(state *transport.State) func(http.ResponseWriter, *http.Request) error {
	return func(w http.ResponseWriter, r *http.Request) error {
		id, err := secretID(r, "id")
		if err != nil {
			return err
		}
		store, secret, err := locate(r, state, transport.Caller(r.Context()), id)
		if err != nil {
			return err
		}
		grants, err := state.Access.GrantsOn(r.Context(), store, secret)
		if err != nil {
			return transport.Internal(err)
		}
		out := make([]GrantView, 0, len(grants))
		for _, grant := range grants {
			out = append(out, viewOf(grant))
		}
		transport.WriteJSON(w, out)
		return nil
	}
}

type delegateBody struct {
	// SubjectKind is `user` or `group`, always explicit — a team called `sre`
	// and a person called `sre` are different subjects (ADR-0003).
	SubjectKind *string  `json:"subject_kind"`
	Subject     *string  `json:"subject"`
	Level       *string  `json:"level"`
	Keys        []string `json:"keys"`
	Days        int64    `json:"days"`
	Note        string   `json:"note"`
}

func delegate(state *transport.State) func(http.ResponseWriter, *http.Request) error {
	return func(w http.ResponseWriter, r *http.Request) error {
		ctx := r.Context()
		actor := transport.Caller(ctx)
		id, err := secretID(r, "id")
		if err != nil {
			return err
		}
		var body delegateBody
		if err := transport.DecodeJSON(r, &body); err != nil {
			return err
		}
		switch {
		case body.SubjectKind == nil:
			return transport.MissingField("subject_kind")
		case body.Subject == nil:
			return transport.MissingField("subject")
		case body.Level == nil:
			return transport.MissingField("level")
		}

		store, secret, err := locate(r, state, actor, id)
		if err != nil {
			return err
		}

		// Only an owner or an admin hands out sight of a secret. A delegation
		// at `write` does not carry the right to re-delegate: that belongs to
		// ownership, which is a different act with a different audit line.
		access, err := state.Access.AccessFor(ctx, actor, store, secret, state.Now())
		if err != nil {
			return transport.Internal(err)
		}
		if access.Basis != entity.BasisOwner && access.Basis != entity.BasisAdmin {
			return transport.Forbidden()
		}

		var subject entity.Subject
		switch *body.SubjectKind {
		case "user":
			subject = entity.User(*body.Subject)
		case "group":
			subject = entity.Group(*body.Subject)
		default:
			return transport.BadRequest(fmt.Sprintf(
				"subject_kind must be user or group, not %q", *body.SubjectKind))
		}
		level, err := entity.ParseLevel(*body.Level)
		if err != nil {
			return transport.BadRequest(err.Error())
		}

		now := state.Now()
		grant := entity.Delegation{
			ID:        uuid.New(),
			Store:     store,
			Secret:    secret,
			Subject:   subject,
			Level:     level,
			Keys:      body.Keys,
			GrantedBy: actor.Handle(),
			GrantedAt: now,
			Note:      body.Note,
		}
		if body.Days > 0 {
			expires := now.Add(time.Duration(body.Days) * 24 * time.Hour)
			grant.ExpiresAt = &expires
		}

		if err := state.Access.Delegate(ctx, grant); err != nil {
			return transport.Internal(err)
		}

		record := auditentity.NewRecord(auditentity.Delegate, id, store, secret).
			WithSubject(*body.Subject).WithKeys(grant.Keys).WithNote(grant.Note)
		if err := state.Audit.Record(ctx, actor, record); err != nil {
			return transport.Internal(err)
		}

		transport.WriteJSON(w, viewOf(grant))
		return nil
	}
}

func revoke(state *transport.State) func(http.ResponseWriter, *http.Request) error {
	return func(w http.ResponseWriter, r *http.Request) error {
		ctx := r.Context()
		actor := transport.Caller(ctx)
		id, err := secretID(r, "id")
		if err != nil {
			return err
		}
		grantID, err := secretID(r, "grant")
		if err != nil {
			return err
		}

		store, secret, err := locate(r, state, actor, id)
		if err != nil {
			return err
		}

		access, err := state.Access.AccessFor(ctx, actor, store, secret, state.Now())
		if err != nil {
			return transport.Internal(err)
		}
		if access.Basis != entity.BasisOwner && access.Basis != entity.BasisAdmin {
			return transport.Forbidden()
		}

		grants, err := state.Access.GrantsOn(ctx, store, secret)
		if err != nil {
			return transport.Internal(err)
		}
		subject := ""
		found := false
		for _, grant := range grants {
			if grant.ID == grantID {
				subject = grant.Subject.ID()
				found = true
				break
			}
		}
		if !found {
			return transport.NotFound()
		}

		removed, err := state.Access.Revoke(ctx, grantID)
		if err != nil {
			return transport.Internal(err)
		}
		if !removed {
			return transport.NotFound()
		}

		record := auditentity.NewRecord(auditentity.Revoke, id, store, secret).
			WithSubject(subject)
		if err := state.Audit.Record(ctx, actor, record); err != nil {
			return transport.Internal(err)
		}

		w.WriteHeader(http.StatusNoContent)
		return nil
	}
}

// secretID reads one uuid path parameter; anything else is 400, the way the
// Rust Path<Uuid> extractor answered.
func secretID(r *http.Request, param string) (uuid.UUID, error) {
	id, err := uuid.Parse(chi.URLParam(r, param))
	if err != nil {
		return uuid.Nil, transport.BadRequest("the id must be a uuid")
	}
	return id, nil
}

// locate is which secret a uuid names, for a caller who can already see it.
//
// The same scan the secrets transport does, duplicated the way the Rust
// domains each carried their own — a shared resolver would couple the two
// domains over three lines of loop.
func locate(
	r *http.Request, state *transport.State, actor identityentity.Actor, id uuid.UUID,
) (string, string, error) {
	ctx := r.Context()
	now := state.Now()
	for _, store := range state.Stores.All() {
		listed, err := store.List(ctx)
		if err != nil {
			continue
		}
		for _, secret := range listed {
			if secretsentity.IDFor(secret) != id {
				continue
			}
			access, err := state.Access.AccessFor(ctx, actor, secret.Store, secret.Name, now)
			if err != nil {
				return "", "", transport.Internal(err)
			}
			if !access.IsVisible() {
				return "", "", transport.NotFound()
			}
			return secret.Store, secret.Name, nil
		}
	}
	return "", "", transport.NotFound()
}
